draft-ietf-ppm-dap.md

```diff
@@ -1492,33 +1531,74 @@ if it should continue preparing the report share:
-
-If successful, the helper responds to the leader with HTTP status 200 OK, media
-type `application/dap-aggregation-job` and a body consisting of an
-`AggregationJob` (see {{aggregation-helper-init}}) compiled from the stepped
-`PrepareStep`s. The order of the steps MUST match that of the `PrepareStep`
-values in the leader's `AggregationJob`. The helper then awaits the next
-message from the leader.
-
-[[OPEN ISSUE: consider relaxing this ordering constraint. See issue#217.]]
+`prep_state` is the helper's current preparation state. This step yields one of
+three outputs:
+
+* An error, in which case the report share is marked as failed and the Helper
+  replies to the Leader with a `PrepareStep` in the `failed` state.
+* An output share, in which case the Helper stores the output share for future
+  collection as described in {{collect-flow}} and replies to the Leader with a
+  `PrepareStep` in the `finished` state.
+* An updated VDAF state and preparation message, denoted
+  `(prep_state, prep_msg)`, in which case the Helper replies to the Leader with
+  a `PrepareStep` in the `continued` state containing `prep_msg`.
+
+After stepping each state, the Helper advances its aggregation job to the
+Leader's `AggregationJobContinueReq.round`.
+
+If the `round` in the Leader's request is equal to the Helper's current round
+(i.e., this is not the first time the Leader has sent this request), then the
+Helper responds with the current round's prepare messages. The Helper SHOULD
+verify that the contents of the `AggregationJobContinueReq` are identical to the
+previous message (see {{aggregation-round-skew-recovery}}).
+
+If the Leader's `round` is behind or more than one round ahead of the Helper's
+current round, then the Helper MUST abort with an error of type `roundMismatch`.
+
+If successful, the Helper responds to the Leader with HTTP status 200 OK, media
+type `application/dap-aggregation-job-resp` and a body consisting of an
+`AggregationJobResp` (see {{aggregation-helper-init}}) compiled from the stepped
+`PrepareStep`s.
+
+#### Recovering From Round Skew {#aggregation-round-skew-recovery}
+
+`AggregationJobContinueReq` messages contain a `round` field, allowing
+Aggregators to ensure that their peer is on an expected round of the VDAF
+preparation algorithm. In particular, the intent is to allow recovery from a
+scenario where the Helper successfully advances from round `n` to `n+1`, but its
+`AggregationJobResp` response to the Leader gets dropped due to something like a
+transient network failure. The Leader could then resend the request to have the
+Helper advance to round `n+1` and the Helper should be able to retransmit the
+`AggregationJobContinueReq` that was previously dropped. To make that kind of
+recovery possible, Aggregator implementations SHOULD checkpoint preparation
+state to durable storage such that Leaders can re-construct continuation
+requests and Helpers can re-construct continuation responses as needed.
```

Comment:
* What can go wrong if an Aggregator doesn't follow this SHOULD? Should this be a MUST?
* What state is "checkpointed"? It would probably be helpful to be specific here: Do we need to put off committing nonces, or output shares, until a certain point?